servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+      mMetaManager.resizeTempBlockMeta(tempBlockMeta, tempBlockMeta.getBlockSize() + moreBytes);
+    } finally {
+      mEvictionLock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public void moveBlock(long userId, long blockId, BlockStoreLocation newLocation)
+      throws IOException {
+    mEvictionLock.readLock().lock();
+    try {
+      long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+      BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+      BlockStoreLocation oldLocation = blockMeta.getBlockLocation();
+      try {
+        moveBlockNoLock(blockId, newLocation);
+        blockMeta = mMetaManager.getBlockMeta(blockId);
+        BlockStoreLocation actualNewLocation = blockMeta.getBlockLocation();
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onMoveBlockByClient(userId, blockId, oldLocation, actualNewLocation);
+        }
+      } finally {
+        mLockManager.unlockBlock(lockId);
+      }
+    } finally {
+      // If we fail to lock, the block is no longer in tiered store
+      mEvictionLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void removeBlock(long userId, long blockId) throws IOException {
+    mEvictionLock.readLock().lock();
+    try {
+      long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+      try {
+        removeBlockNoLock(userId, blockId);
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onRemoveBlockByClient(userId, blockId);
+        }
+      } finally {
+        mLockManager.unlockBlock(lockId);
+      }
+    } finally {
+      // If we fail to lock, the block is no longer in tiered store
+      mEvictionLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void accessBlock(long userId, long blockId) {
+    for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+      listener.onAccessBlock(userId, blockId);
+    }
+  }
+
+  @Override
+  public void freeSpace(long userId, long availableBytes, BlockStoreLocation location)
+      throws IOException {
+    mEvictionLock.writeLock().lock();
+    try {
+      freeSpaceInternal(userId, availableBytes, location);
+    } finally {
+      mEvictionLock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public void cleanupUser(long userId) throws IOException {
+    mEvictionLock.readLock().lock();
+    List<TempBlockMeta> tempBlocksToRemove = mMetaManager.cleanupUser(userId);
+    mLockManager.cleanupUser(userId);
+    mEvictionLock.readLock().unlock();
+    Set<String> dirs = new HashSet<String>();
+    for (TempBlockMeta tempBlockMeta : tempBlocksToRemove) {
+      String fileName = tempBlockMeta.getPath();
+      try {
+        String dirName = CommonUtils.getParent(fileName);
+        dirs.add(dirName);
+      } catch (InvalidPathException e) {
+        LOG.error("Cannot parse parent dir of {}", fileName);
+      }
+      if (!new File(fileName).delete()) {
```

Comment:
I think the current code has potential risk condition when cleaning up blocks. I left a TODO and let's improve it in later PR.